xarray/tests/__init__.py

```diff
@@ -234,6 +236,129 @@ def __getitem__(self, key):
+def broadcast_to(
+    x: ConcatenatableArray, /, shape: tuple[int, ...]
+) -> ConcatenatableArray:
+    """
+    Broadcasts an array to a specified shape, by either manipulating chunk keys or copying chunk manifest entries.
+    """
+    if not isinstance(x, ConcatenatableArray):
+        raise TypeError
+
+    result = np.broadcast_to(x._array, shape=shape)
+    return ConcatenatableArray(result)
+
+
+class ConcatenatableArray:
+    """Disallows loading or coercing to an index but does support concatenation / stacking."""
+
+    HANDLED_ARRAY_FUNCTIONS = [concatenate, stack, result_type]
+
+    def __init__(self, array):
+        # use ._array instead of .array because we don't want this to be accessible even to xarray's internals (e.g. create_default_index_implicit)
+        self._array = array
+
+    @property
+    def dtype(self: Any) -> np.dtype:
+        return self._array.dtype
+
+    @property
+    def shape(self: Any) -> tuple[int, ...]:
+        return self._array.shape
+
+    @property
+    def ndim(self: Any) -> int:
+        return self._array.ndim
+
+    def __repr__(self: Any) -> str:
+        return f"{type(self).__name__}(array={self._array!r})"
+
+    def get_duck_array(self):
+        raise UnexpectedDataAccess("Tried accessing data")
+
+    def __array__(self, dtype: np.typing.DTypeLike = None):
+        raise UnexpectedDataAccess("Tried accessing data")
+
+    def __getitem__(self, key) -> ConcatenatableArray:
+        """Some cases of concat require supporting expanding dims by dimensions of size 1"""
+        # see https://data-apis.org/array-api/2022.12/API_specification/indexing.html#multi-axis-indexing
+        arr = self._array
+        for axis, indexer_1d in enumerate(key):
+            if indexer_1d is None:
+                arr = np.expand_dims(arr, axis)
+            elif indexer_1d is Ellipsis:
+                pass
+            else:
+                raise UnexpectedDataAccess("Tried accessing data.")
+        return ConcatenatableArray(arr)
+
+    def __array_function__(self, func, types, args, kwargs) -> Any:
+        if func not in HANDLED_ARRAY_FUNCTIONS:
+            return NotImplemented
+
+        # Note: this allows subclasses that don't override
+        # __array_function__ to handle ManifestArray objects
+        if not all(issubclass(t, ConcatenatableArray) for t in types):
+            return NotImplemented
+
+        return HANDLED_ARRAY_FUNCTIONS[func](*args, **kwargs)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs) -> Any:
+        """We have to define this in order to convince xarray that this class is a duckarray, even though we will never support ufuncs."""
+        return NotImplemented
+
+    def astype(self, dtype: np.dtype, /, *, copy: bool = True) -> ConcatenatableArray:
+        """Needed because xarray will call this even when it's a no-op"""
+        if dtype != self.dtype:
+            raise NotImplementedError()
+        else:
+            return self
+
+
```

Comment:
should we move this to a separate module (which we can import in `xarray.tests`). `__init__` is getting a bit big

Reply:
Moved to a new `arrays.py` module in https://github.com/pydata/xarray/pull/8872/commits/6d825e5d97088037933c3242d136b9c41488fd07